usr.bin/perl5/ext/SDBM_File/sdbm.3
.\" $Id: sdbm.3,v 1.2 90/12/13 13:00:57 oz Exp $
.TH SDBM 3 "1 March 1990"
.SH NAME
sdbm, sdbm_open, sdbm_prep, sdbm_close, sdbm_fetch, sdbm_store, sdbm_delete, sdbm_exists, sdbm_firstkey, sdbm_nextkey, sdbm_hash, sdbm_rdonly, sdbm_error, sdbm_clearerr, sdbm_dirfno, sdbm_pagfno \- data base subroutines
.SH SYNOPSIS
.nf
.ft B
#include <sdbm.h>
.sp
typedef struct {
	char *dptr;
	int dsize;
} datum;
.sp
datum nullitem = { NULL, 0 };
.sp
\s-1DBM\s0 *sdbm_open(char *file, int flags, int mode)
.sp
\s-1DBM\s0 *sdbm_prep(char *dirname, char *pagname, int flags, int mode)
.sp
void sdbm_close(\s-1DBM\s0 *db)
.sp
datum sdbm_fetch(\s-1DBM\s0 *db, key)
.sp
int sdbm_store(\s-1DBM\s0 *db, datum key, datum val, int flags)
.sp
int sdbm_delete(\s-1DBM\s0 *db, datum key)
.sp
int sdbm_exists(\s-1DBM\s0 *db, datum key)
.sp
datum sdbm_firstkey(\s-1DBM\s0 *db)
.sp
datum sdbm_nextkey(\s-1DBM\s0 *db)
.sp
long sdbm_hash(char *string, int len)
.sp
int sdbm_rdonly(\s-1DBM\s0 *db)
int sdbm_error(\s-1DBM\s0 *db)
sdbm_clearerr(\s-1DBM\s0 *db)
int sdbm_dirfno(\s-1DBM\s0 *db)
int sdbm_pagfno(\s-1DBM\s0 *db)
.ft R
.fi
.SH DESCRIPTION
.IX "database library" sdbm "" "\fLsdbm\fR"
.IX sdbm_open "" "\fLsdbm_open\fR \(em open \fLsdbm\fR database"
.IX sdbm_prep "" "\fLsdbm_prep\fR \(em prepare \fLsdbm\fR database"
.IX sdbm_close "" "\fLsdbm_close\fR \(em close \fLsdbm\fR routine"
.IX sdbm_fetch "" "\fLsdbm_fetch\fR \(em fetch \fLsdbm\fR database data"
.IX sdbm_store "" "\fLsdbm_store\fR \(em add data to \fLsdbm\fR database"
.IX sdbm_delete "" "\fLsdbm_delete\fR \(em remove data from \fLsdbm\fR database"
.IX sdbm_exists "" "\fLsdbm_exists\fR \(em test \fLsdbm\fR key existence"
.IX sdbm_firstkey "" "\fLsdbm_firstkey\fR \(em access \fLsdbm\fR database"
.IX sdbm_nextkey "" "\fLsdbm_nextkey\fR \(em access \fLsdbm\fR database"
.IX sdbm_hash "" "\fLsdbm_hash\fR \(em string hash for \fLsdbm\fR database"
.IX sdbm_rdonly "" "\fLsdbm_rdonly\fR \(em return \fLsdbm\fR database read-only mode"
.IX sdbm_error "" "\fLsdbm_error\fR \(em return \fLsdbm\fR database error condition"
.IX sdbm_clearerr "" "\fLsdbm_clearerr\fR \(em clear \fLsdbm\fR database error condition"
.IX sdbm_dirfno "" "\fLsdbm_dirfno\fR \(em return \fLsdbm\fR database bitmap file descriptor"
.IX sdbm_pagfno "" "\fLsdbm_pagfno\fR \(em return \fLsdbm\fR database data file descriptor"
.IX "database functions \(em \fLsdbm\fR"  sdbm_open  ""  \fLsdbm_open\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_prep  ""  \fLsdbm_prep\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_close  ""  \fLsdbm_close\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_fetch  ""  \fLsdbm_fetch\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_store  ""  \fLsdbm_store\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_delete  ""  \fLsdbm_delete\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_firstkey  ""  \fLsdbm_firstkey\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_nextkey  ""  \fLsdbm_nextkey\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_rdonly  ""  \fLsdbm_rdonly\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_error  ""  \fLsdbm_error\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_clearerr  ""  \fLsdbm_clearerr\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_dirfno  ""  \fLsdbm_dirfno\fP
.IX "database functions \(em \fLsdbm\fR"  sdbm_pagfno  ""  \fLsdbm_pagfno\fP
.LP
This package allows an application to maintain a mapping of <key,value> pairs
in disk files.  This is not to be considered a real database system, but is
still useful in many simple applications built around fast retrieval of a data
value from a key.  This implementation uses an external hashing scheme,
called Dynamic Hashing, as described by Per-Aake Larson in BIT 18 (1978) pp.
184-201.  Retrieval of any item usually requires a single disk access.
The application interface is compatible with the
.IR ndbm (3)
library.
.LP
An
.B sdbm
database is kept in two files usually given the extensions
.B \.dir
and
.BR \.pag .
The
.B \.dir
file contains a bitmap representing a forest of binary hash trees, the leaves
of which indicate data pages in the
.B \.pag
file.
.LP
The application interface uses the
.B datum
structure to describe both
.I keys
and
.IR value s.
A
.B datum
specifies a byte sequence of
.I dsize
size pointed to by
.IR dptr .
If you use
.SM ASCII
strings as
.IR key s
or
.IR value s,
then you must decide whether or not to include the terminating
.SM NUL
byte which sometimes defines strings.  Including it will require larger
database files, but it will be possible to get sensible output from a
.IR strings (1)
command applied to the data file.
.LP
In order to allow a process using this package to manipulate multiple
databases, the applications interface always requires a
.IR handle ,
a
.BR "DBM *" ,
to identify the database to be manipulated.  Such a handle can be obtained
from the only routines that do not require it, namely
.BR sdbm_open (\|)
or
.BR sdbm_prep (\|).
Either of these will open or create the two necessary files.  The
difference is that the latter allows explicitly naming the bitmap and data
files whereas
.BR sdbm_open (\|)
will take a base file name and call
.BR sdbm_prep (\|)
with the default extensions.
The
.I flags
and
.I mode
parameters are the same as for
.BR open (2).
.LP
To free the resources occupied while a database handle is active, call
.BR sdbm_close (\|).
.LP
Given a handle, one can retrieve data associated with a key by using the
.BR sdbm_fetch (\|)
routine, and associate data with a key by using the
.BR sdbm_store (\|)
routine.
.BR sdbm_exists (\|)
will say whether a given key exists in the database.
.LP
The values of the
.I flags
parameter for
.BR sdbm_store (\|)
can be either
.BR \s-1DBM_INSERT\s0 ,
which will not change an existing entry with the same key, or
.BR \s-1DBM_REPLACE\s0 ,
which will replace an existing entry with the same key.
Keys are unique within the database.
.LP
To delete a key and its associated value use the
.BR sdbm_delete (\|)
routine.
.LP
To retrieve every key in the database, use a loop like:
.sp
.nf
.ft B
for (key = sdbm_firstkey(db); key.dptr != NULL; key = sdbm_nextkey(db))
        ;
.ft R
.fi
.LP
The order of retrieval is unspecified.
.LP
If you determine that the performance of the database is inadequate or
you notice clustering or other effects that may be due to the hashing
algorithm used by this package, you can override it by supplying your
own
.BR sdbm_hash (\|)
routine.  Doing so will make the database unintelligible to any other
applications that do not use your specialized hash function.
.sp
.LP
The following macros are defined in the header file:
.IP
.BR sdbm_rdonly (\|)
returns true if the database has been opened read\-only.
.IP
.BR sdbm_error (\|)
returns true if an I/O error has occurred.
.IP
.BR sdbm_clearerr (\|)
allows you to clear the error flag if you think you know what the error
was and insist on ignoring it.
.IP
.BR sdbm_dirfno (\|)
returns the file descriptor associated with the bitmap file.
.IP
.BR sdbm_pagfno (\|)
returns the file descriptor associated with the data file.
.SH SEE ALSO
.IR open (2).
.SH DIAGNOSTICS
Functions that return a
.B "DBM *"
handle will use
.SM NULL
to indicate an error.
Functions that return an
.B int
will use \-1 to indicate an error.  The normal return value in that case is 0.
Functions that return a
.B datum
will return
.B nullitem
to indicate an error.
.LP
As a special case of
.BR sdbm_store (\|),
if it is called with the
.B \s-1DBM_INSERT\s0
flag and the key already exists in the database, the return value will be 1.
.LP
In general, if a function parameter is invalid,
.B errno
will be set to
.BR \s-1EINVAL\s0 .
If a write operation is requested on a read-only database,
.B errno
will be set to
.BR \s-1ENOPERM\s0 .
If a memory allocation (using
.IR malloc (3))
failed,
.B errno
will be set to
.BR \s-1ENOMEM\s0 .
For I/O operation failures
.B errno
will contain the value set by the relevant failed system call, either
.IR read (2),
.IR write (2),
or
.IR lseek (2).
.SH AUTHOR
.IP "Ozan S. Yigit" (oz@nexus.yorku.ca)
.SH BUGS
The sum of key and value data sizes must not exceed
.B \s-1PAIRMAX\s0
(1008 bytes).
.LP
The sum of the key and value data sizes where several keys hash to the
same value must fit within one bitmap page.
.LP
The
.B \.pag
file will contain holes, so its apparent size is larger than its contents.
When copied through the filesystem the holes will be filled.
.LP
The contents of
.B datum
values returned are in volatile storage.  If you want to retain the values
pointed to, you must copy them immediately before another call to this package.
.LP
The only safe way for multiple processes to (read and) update a database at
the same time, is to implement a private locking scheme outside this package
and open and close the database between lock acquisitions.  It is safe for
multiple processes to concurrently access a database read-only.
.SH APPLICATIONS PORTABILITY
For complete source code compatibility with the Berkeley Unix
.IR ndbm (3)
library, the 
.B sdbm.h
header file should be installed in
.BR /usr/include/ndbm.h .
.LP
The
.B nullitem
data item, and the
.BR sdbm_prep (\|),
.BR sdbm_hash (\|),
.BR sdbm_rdonly (\|),
.BR sdbm_dirfno (\|),
and
.BR sdbm_pagfno (\|)
functions are unique to this package.


libs-gui/NEWS
1 NEWS
******

The currently released version of the library is '0.28.0'.

1.1 Noteworthy changes in version '0.28.0'
==========================================

This version adds support for modern XIB files and many new classes.
Plus the usual bunch of bug fixes.

   * Support loading of document XIB files.
   * Improve Key Value Binding for NSArrayController and add more
     bindings.
   * Better support for multi monitor usage and other improvement in the
     backend integration.
   * Add classes NSFontCollection, NSColorSampler, NSSpeechRecognizer,
     NSAppearance, NSPDFInfo, NSPICTImageRep, NSCIImageRep,
     NSPDFImageRep, NSPDFPanel, NSDataAsset, NSDatePicker,
     NSDatePickerCell, NSPredicateEditor, NSPredicateEditorRowTemplate,
     NSRuleEditor, NSGestureRecognizer, NSButtonTouchBarItem,
     NSCandidateListTouchBarItem, NSClickGestureRecognizer,
     NSColorPickerTouchBarItem, NSCustomTouchBarItem,
     NSGroupTouchBarItem, NSMagnificationGestureRecognizer,
     NSPanGestureRecognizer, NSPickerTouchBarItem,
     NSPopoverTouchBarItem, NSPressGestureRecognizer,
     NSRotationGestureRecognizer, NSSharingServicePickerTouchBarItem,
     NSSliderTouchBarItem, NSStepperTouchBarItem, NSTouchBarItem,
     NSTouchBar, NSTouch, NSDockTile.
   * Implement NSEPSImageRep.
   * Better encoding handling in RTF files.
   * Theming and drawing improvements.
   * Increase small font size to 10.
   * New cursor and stepper images.
   * Move NSFileWrapper to Foundation.
   * Fixed build on Debian GNU/kFreeBSD.
   * With command line argument -autolaunch YES, do not activate the
     application when -activateIgnoringOtherApps: is invoked.
   * Improvements to WindowMaker compatibility (e.g.  WMFHideApplication
     support).
   * Lowered NSFloatingWindowLevel by one to distinguish floating panels
     from menus.

1.2 Noteworthy changes in version '0.27.0'
==========================================

This version includes numerous bugfixes, compatibility improvements and
other changes accumulated over the last year.  It also enables work to
be done on integrating NSViews with a Core Animation renderer.

   * Make targetForAction safer.
   * Speed up menu updates.
   * Clean up speech tool compilation and switch to newer interface.
   * Fix bug in CUPS subclassing introduced in last release.
   * Minor improvements to typesetting.
   * Add NSIsControllerMarker.
   * Fix tracking on segmented cell.
   * Bring slider cell closer to Cocoa implementation.
   * Add ivar for Core Animation in NSView.
   * Improve border calculation on printing.
   * Lazy load app icon.
   * Better detection of removable volumes.
   * Polish translations.
   * Japanese translations.
   * Lots of bug fixes.

1.3 Noteworthy changes in version '0.26.2'
==========================================

This version is a small, but important bugfix release.

   * printing: Fix allocation of the CUPS printing classes.
   * installation: Fix the configure script.

1.4 Noteworthy changes in version '0.26.1'
==========================================

This version is released to conincide with version 1.25.1 of
gnustep-base, which contains changes required for this version of
gnustep-gui and gnustep-back.

   It includes an important workaround for users of GNUstep Objective-C
Runtime (libobjc2) and non-fragile ABI to avoid a bug in interaction
between the clang compiler and the runtime when non-fragile ABI is in
use.  Specifically, Clang and the runtime may disagree on what is the
offset of an ivar in a class's RAM. This manifested in a crash at
application startup due to misalignment of _gcontext inside NSThread.
See the mailing list discussion
(http://lists.gnu.org/archive/html/discuss-gnustep/2017-12/msg00129.html)
for more information.

   It also contains the following changes:

   * tests: Cleanup of warnings.
   * tests: Fix text system deallocation test.
   * printing: Undefine __BLOCKS__ before including cups.h, as some
     versions of the header expect that libdispatch is present and used
     if __BLOCKS__ is defined.
   * graphics context: Workaround for Clang+libobjc2+nonfragile ABI
     issue.

1.5 Noteworthy changes in version '0.26.0'
==========================================

This version was bumped due to previous binary incompatibilities between
0.25.0 and 0.25.1.  This version also includes numerous compatibility
improvements from the Summer of Code project, and a wide variety of
other fixes.  Notably, it fixes the use of cupsGetPPD() in the printing
system.

   * printing: Add an include to get deprecated function cupsGetPPD() on
     newer CUPS systems.
   * chore: Bump required base version.
   * tiff: Support for writing resolution.
   * jpeg: Save resolution information if it is different from 72 dpi.
   * save panel: Fix return type of sorting function.
   * events: Add some newer Cocoa enums and one method with dummy
     implementation.
   * speech synthesis: NSSpeechSynthesizerDelegate is now a @protocol on
     runtimes that support it.
   * pasteboard: New type identifiers.
   * translations: Some work on Polish, Russian and German translations
   * cell: Improvements to mouse tracking logic on NSCell.
   * image: If an unknown named image is unarchived with a coder or
     keyed coder, keep the name.
   * screen: Add -backingScaleFactor and return 1.0.
   * window: Return 1.0 from -backingScaleFactor.

   * compatibility: Numerous stub implementations of constants, classes
     and methods to improve source-level compatibility.
   * other bugfixes

1.6 Noteworthy changes in version '0.25.1'
==========================================

   * JPEG (saving) alpha channel fixes and size with resolution != 72
   * JPEG resolution read support
   * TIFF saving fixes
   * Improved volumes mounting and support
   * Portability improvements in volume mounting and support
   * Corrected layout of empty strings
   * Only update visible menus

1.7 Noteworthy changes in version '0.25.0'

usr.bin/perl5/ext/SDBM_File/sdbm.c
/*
 * sdbm - ndbm work-alike hashed database library
 * based on Per-Aake Larson's Dynamic Hashing algorithms. BIT 18 (1978).
 * author: oz@nexus.yorku.ca
 * status: public domain.
 *
 * core routines
 */

#include "config.h"
#ifdef WIN32
#include "io.h"
#endif
#include "sdbm.h"
#include "tune.h"
#include "pair.h"

#ifdef I_FCNTL
# include <fcntl.h>
#endif
#ifdef I_SYS_FILE
# include <sys/file.h>
#endif

#include <string.h>

/*
 * externals
 */

#include <errno.h> /* See notes in perl.h about avoiding
                        extern int errno; */
#ifdef __cplusplus
extern "C" {
#endif

extern Malloc_t malloc(MEM_SIZE);
extern Free_t free(Malloc_t);

#ifdef __cplusplus
}
#endif

const datum nullitem = {0, 0};

#ifdef WIN32
#  undef lseek
#  define lseek _lseeki64
#endif

/*
 * forward
 */
static int getdbit(DBM *, long);
static int setdbit(DBM *, long);
static int getpage(DBM *, long);
static datum getnext(DBM *);
static int makroom(DBM *, long, int);

/*
 * useful macros
 */
#define bad(x)		((x).dptr == NULL || (x).dsize < 0)
#define exhash(item)	sdbm_hash((item).dptr, (item).dsize)
#define ioerr(db)	((db)->flags |= DBM_IOERR)

#define OFF_PAG(off)	(Off_t) (off) * PBLKSIZ
#define OFF_DIR(off)	(Off_t) (off) * DBLKSIZ

static const long masks[] = {
        000000000000, 000000000001, 000000000003, 000000000007,
        000000000017, 000000000037, 000000000077, 000000000177,
        000000000377, 000000000777, 000000001777, 000000003777,
        000000007777, 000000017777, 000000037777, 000000077777,
        000000177777, 000000377777, 000000777777, 000001777777,
        000003777777, 000007777777, 000017777777, 000037777777,
        000077777777, 000177777777, 000377777777, 000777777777,
        001777777777, 003777777777, 007777777777, 017777777777
};

DBM *
sdbm_open(char *file, int flags, int mode)
{
        DBM *db;
        char *dirname;
        char *pagname;
        size_t filelen;
        const size_t dirfext_size = sizeof(DIRFEXT "");
        const size_t pagfext_size = sizeof(PAGFEXT "");

        if (file == NULL || !*file)
                return errno = EINVAL, (DBM *) NULL;
/*
 * need space for two separate filenames
 */
        filelen = strlen(file);

        if ((dirname = (char *) malloc(filelen + dirfext_size
                                       + filelen + pagfext_size)) == NULL)
                return errno = ENOMEM, (DBM *) NULL;
/*
 * build the file names
 */
        memcpy(dirname, file, filelen);
        memcpy(dirname + filelen, DIRFEXT, dirfext_size);
        pagname = dirname + filelen + dirfext_size;
        memcpy(pagname, file, filelen);
        memcpy(pagname + filelen, PAGFEXT, pagfext_size);

        db = sdbm_prep(dirname, pagname, flags, mode);
        free((char *) dirname);
        return db;
}

DBM *
sdbm_prep(char *dirname, char *pagname, int flags, int mode)
{
        DBM *db;
        struct stat dstat;

        if ((db = (DBM *) malloc(sizeof(DBM))) == NULL)
                return errno = ENOMEM, (DBM *) NULL;

        db->flags = 0;
        db->hmask = 0;
        db->blkptr = 0;
        db->keyptr = 0;
/*
 * adjust user flags so that WRONLY becomes RDWR, 
 * as required by this package. Also set our internal
 * flag for RDONLY if needed.
 */
        if (flags & O_WRONLY)
                flags = (flags & ~O_WRONLY) | O_RDWR;

        else if ((flags & 03) == O_RDONLY)
                db->flags = DBM_RDONLY;
/*
 * open the files in sequence, and stat the dirfile.
 * If we fail anywhere, undo everything, return NULL.
 */
#if defined(OS2) || defined(WIN32) || defined(__CYGWIN__)
        flags |= O_BINARY;
#	endif
        if ((db->pagf = open(pagname, flags, mode)) > -1) {
                if ((db->dirf = open(dirname, flags, mode)) > -1) {
/*
 * need the dirfile size to establish max bit number.
 */
                        if (fstat(db->dirf, &dstat) == 0) {
/*
 * zero size: either a fresh database, or one with a single,
 * unsplit data page: dirpage is all zeros.
 */
                                db->dirbno = (!dstat.st_size) ? 0 : -1;
                                db->pagbno = -1;
                                db->maxbno = dstat.st_size * BYTESIZ;

                                (void) memset(db->pagbuf, 0, PBLKSIZ);
                                (void) memset(db->dirbuf, 0, DBLKSIZ);
                        /*
                         * success
                         */
                                return db;
                        }
                        (void) close(db->dirf);
                }
                (void) close(db->pagf);
        }
        free((char *) db);
        return (DBM *) NULL;
}

void
sdbm_close(DBM *db)
{
        if (db == NULL)
                errno = EINVAL;
        else {
                (void) close(db->dirf);
                (void) close(db->pagf);
                free((char *) db);
        }
}

datum
sdbm_fetch(DBM *db, datum key)
{
        if (db == NULL || bad(key))
                return errno = EINVAL, nullitem;

        if (getpage(db, exhash(key)))
                return getpair(db->pagbuf, key);

        return ioerr(db), nullitem;
}

int
sdbm_exists(DBM *db, datum key)
{
        if (db == NULL || bad(key))
                return errno = EINVAL, -1;

        if (getpage(db, exhash(key)))
                return exipair(db->pagbuf, key);

        return ioerr(db), -1;
}

int
sdbm_delete(DBM *db, datum key)
{
        if (db == NULL || bad(key))
                return errno = EINVAL, -1;
        if (sdbm_rdonly(db))
                return errno = EPERM, -1;

        if (getpage(db, exhash(key))) {
                if (!delpair(db->pagbuf, key))
                        return -1;
/*
 * update the page file
 */
                if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
                    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
                        return ioerr(db), -1;

                return 0;
        }

        return ioerr(db), -1;
}

int
sdbm_store(DBM *db, datum key, datum val, int flags)
{
        int need;
        long hash;

        if (db == NULL || bad(key))
                return errno = EINVAL, -1;
        if (sdbm_rdonly(db))
                return errno = EPERM, -1;

        need = key.dsize + val.dsize;
/*
 * is the pair too big (or too small) for this database ??
 */
        if (need < 0 || need > PAIRMAX)
                return errno = EINVAL, -1;

        if (getpage(db, (hash = exhash(key)))) {
/*
 * if we need to replace, delete the key/data pair
 * first. If it is not there, ignore.
 */
                if (flags == DBM_REPLACE)
                        (void) delpair(db->pagbuf, key);
#ifdef SEEDUPS
                else if (duppair(db->pagbuf, key))
                        return 1;
#endif
/*
 * if we do not have enough room, we have to split.
 */
                if (!fitpair(db->pagbuf, need))
                        if (!makroom(db, hash, need))
                                return ioerr(db), -1;
/*
 * we have enough room or split is successful. insert the key,
 * and update the page file.
 */
                (void) putpair(db->pagbuf, key, val);

                if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
                    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
                        return ioerr(db), -1;
        /*
         * success
         */
                return 0;
        }

        return ioerr(db), -1;
}

/*
 * makroom - make room by splitting the overfull page
 * this routine will attempt to make room for SPLTMAX times before
 * giving up.
 */
static int
makroom(DBM *db, long int hash, int need)
{
        long newp;
        char twin[PBLKSIZ];
#if defined(DOSISH) || defined(WIN32)
        char zer[PBLKSIZ];
        Off_t oldtail;
#endif
        char *pag = db->pagbuf;
        char *New = twin;
        int smax = SPLTMAX;
#ifdef BADMESS
        int rc;
#endif

        do {
/*
 * split the current page
 */
                (void) splpage(pag, New, db->hmask + 1);
/*
 * address of the new page
 */
                newp = (hash & db->hmask) | (db->hmask + 1);

/*
 * write delay, read avoidance/cache shuffle:
 * select the page for incoming pair: if key is to go to the new page,
 * write out the previous one, and copy the new one over, thus making
 * it the current page. If not, simply write the new page, and we are
 * still looking at the page of interest. current page is not updated
 * here, as sdbm_store will do so, after it inserts the incoming pair.
 */

#if defined(DOSISH) || defined(WIN32)
                /*
                 * Fill hole with 0 if made it.
                 * (hole is NOT read as 0)
                 */
                oldtail = lseek(db->pagf, 0L, SEEK_END);
                memset(zer, 0, PBLKSIZ);
                while (OFF_PAG(newp) > oldtail) {
                        if (lseek(db->pagf, 0L, SEEK_END) < 0 ||
                            write(db->pagf, zer, PBLKSIZ) < 0) {

                                return 0;
                        }
                        oldtail += PBLKSIZ;
                }
#endif
                if (hash & (db->hmask + 1)) {
                        if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
                            || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
                                return 0;
                        db->pagbno = newp;
                        (void) memcpy(pag, New, PBLKSIZ);
                }
                else if (lseek(db->pagf, OFF_PAG(newp), SEEK_SET) < 0
                         || write(db->pagf, New, PBLKSIZ) < 0)
                        return 0;

                if (!setdbit(db, db->curbit))
                        return 0;
/*
 * see if we have enough room now
 */
                if (fitpair(pag, need))
                        return 1;
/*
 * try again... update curbit and hmask as getpage would have
 * done. because of our update of the current page, we do not
 * need to read in anything. BUT we have to write the current
 * [deferred] page out, as the window of failure is too great.
 */
                db->curbit = 2 * db->curbit +
                        ((hash & (db->hmask + 1)) ? 2 : 1);
                db->hmask |= db->hmask + 1;

                if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
                    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
                        return 0;

        } while (--smax);
/*
 * if we are here, this is real bad news. After SPLTMAX splits,
 * we still cannot fit the key. say goodnight.
 */
#ifdef BADMESS
        rc = write(2, "sdbm: cannot insert after SPLTMAX attempts.\n", 44);
        /* PERL_UNUSED_VAR() or PERL_UNUSED_RESULT() would be
         * useful here but that would mean pulling in perl.h */
        (void)rc;
#endif
        return 0;

}

/*
 * the following two routines will break if
 * deletions aren't taken into account. (ndbm bug)
 */
datum
sdbm_firstkey(DBM *db)
{
        if (db == NULL)
                return errno = EINVAL, nullitem;
/*
 * start at page 0
 */
        if (lseek(db->pagf, OFF_PAG(0), SEEK_SET) < 0
            || read(db->pagf, db->pagbuf, PBLKSIZ) < 0)
                return ioerr(db), nullitem;
        if (!chkpage(db->pagbuf)) {
            errno = EINVAL;
            ioerr(db);
            db->pagbno = -1;
            return nullitem;
        }
        db->pagbno = 0;
        db->blkptr = 0;
        db->keyptr = 0;

        return getnext(db);
}

datum
sdbm_nextkey(DBM *db)
{
        if (db == NULL)
                return errno = EINVAL, nullitem;
        return getnext(db);
}

/*
 * all important binary trie traversal
 */
static int
getpage(DBM *db, long int hash)
{
        int hbit;
        long dbit;
        long pagb;

        dbit = 0;
        hbit = 0;
        while (dbit < db->maxbno && getdbit(db, dbit))
                dbit = 2 * dbit + ((hash & (1 << hbit++)) ? 2 : 1);

        debug(("dbit: %d...", dbit));

        db->curbit = dbit;
        db->hmask = masks[hbit];

        pagb = hash & db->hmask;
/*
 * see if the block we need is already in memory.
 * note: this lookaside cache has about 10% hit rate.
 */
        if (pagb != db->pagbno) { 
/*
 * note: here, we assume a "hole" is read as 0s.
 * if not, must zero pagbuf first.
 */
                if (lseek(db->pagf, OFF_PAG(pagb), SEEK_SET) < 0
                    || read(db->pagf, db->pagbuf, PBLKSIZ) < 0)
                        return 0;
                if (!chkpage(db->pagbuf)) {
                    errno = EINVAL;
                    db->pagbno = -1;
                    ioerr(db);
                    return 0;
                }
                db->pagbno = pagb;

                debug(("pag read: %d\n", pagb));
        }
        return 1;
}

static int
getdbit(DBM *db, long int dbit)
{
        long c;
        long dirb;

        c = dbit / BYTESIZ;
        dirb = c / DBLKSIZ;

        if (dirb != db->dirbno) {
                int got;
                if (lseek(db->dirf, OFF_DIR(dirb), SEEK_SET) < 0
                    || (got=read(db->dirf, db->dirbuf, DBLKSIZ)) < 0)
                        return 0;
                if (got==0) 
                        memset(db->dirbuf,0,DBLKSIZ);
                db->dirbno = dirb;

                debug(("dir read: %d\n", dirb));
        }

        return db->dirbuf[c % DBLKSIZ] & (1 << dbit % BYTESIZ);
}

static int
setdbit(DBM *db, long int dbit)
{
        long c;
        long dirb;

        c = dbit / BYTESIZ;
        dirb = c / DBLKSIZ;

        if (dirb != db->dirbno) {
                int got;
                if (lseek(db->dirf, OFF_DIR(dirb), SEEK_SET) < 0
                    || (got=read(db->dirf, db->dirbuf, DBLKSIZ)) < 0)
                        return 0;
                if (got==0) 
                        memset(db->dirbuf,0,DBLKSIZ);
                db->dirbno = dirb;

                debug(("dir read: %d\n", dirb));
        }

        db->dirbuf[c % DBLKSIZ] |= (1 << dbit % BYTESIZ);

#if 0
        if (dbit >= db->maxbno)
                db->maxbno += DBLKSIZ * BYTESIZ;
#else
        if (OFF_DIR((dirb+1))*BYTESIZ > db->maxbno) 
                db->maxbno=OFF_DIR((dirb+1))*BYTESIZ;
#endif

        if (lseek(db->dirf, OFF_DIR(dirb), SEEK_SET) < 0
            || write(db->dirf, db->dirbuf, DBLKSIZ) < 0)
                return 0;

        return 1;
}

/*
 * getnext - get the next key in the page, and if done with
 * the page, try the next page in sequence
 */
static datum
getnext(DBM *db)
{
        datum key;

        for (;;) {
                db->keyptr++;
                key = getnkey(db->pagbuf, db->keyptr);
                if (key.dptr != NULL)
                        return key;
/*
 * we either run out, or there is nothing on this page..
 * try the next one... If we lost our position on the
 * file, we will have to seek.
 */
                db->keyptr = 0;
                if (db->pagbno != db->blkptr++)
                        if (lseek(db->pagf, OFF_PAG(db->blkptr), SEEK_SET) < 0)
                                break;
                db->pagbno = db->blkptr;
                if (read(db->pagf, db->pagbuf, PBLKSIZ) <= 0)
                        break;
                if (!chkpage(db->pagbuf)) {
                    errno = EINVAL;
                    db->pagbno = -1;
                    ioerr(db);
                    break;
                }
        }

        return ioerr(db), nullitem;
}

